mesh/meshdb.go

```diff
@@ -145,109 +159,188 @@ func (m *DB) Close() {
-		return ErrAlreadyExist
+// LayerBallots retrieves all ballots from a layer by layer ID.
+func (m *DB) LayerBallots(layerID types.LayerID) ([]*types.Ballot, error) {
+	proposals, err := m.LayerProposals(layerID)
+	if err != nil {
+		return nil, fmt.Errorf("layer ballots: %w", err)
 	}
-	if err := m.writeBlock(bl); err != nil {
-		return fmt.Errorf("write block: %w", err)
+	ballots := make([]*types.Ballot, 0, len(proposals))
+	for _, p := range proposals {
+		ballots = append(ballots, &p.Ballot)
 	}
+	return ballots, nil
+}
 
-	return nil
+// AddBlock adds a block to the database.
+func (m *DB) AddBlock(b *types.Block) error {
+	p := (*types.Proposal)(b)
+	return m.AddProposal(p)
 }
 
-// GetBlock gets a block from the database by id.
-func (m *DB) GetBlock(id types.BlockID) (*types.Block, error) {
-	if blkh := m.blockCache.Get(id); blkh != nil {
+// GetBlock returns the block specified by the block ID.
+func (m *DB) GetBlock(bid types.BlockID) (*types.Block, error) {
+	if blkh := m.blockCache.Get(bid); blkh != nil {
 		return blkh, nil
 	}
 
-	b, err := m.getBlockBytes(id)
+	p, err := m.getProposal(types.ProposalID(bid))
+	if err != nil {
+		return nil, err
+	}
+	return (*types.Block)(p), nil
+}
+
+// LayerBlockIds retrieves all block IDs from the layer specified by layer ID.
+func (m *DB) LayerBlockIds(layerID types.LayerID) ([]types.BlockID, error) {
+	pids, err := m.LayerProposalIDs(layerID)
 	if err != nil {
 		return nil, err
 	}
-	mbk := &types.DBBlock{}
-	if err := types.BytesToInterface(b, mbk); err != nil {
-		return nil, fmt.Errorf("parse block: %w", err)
+	bids := make([]types.BlockID, 0, len(pids))
+	for _, pid := range pids {
+		bids = append(bids, types.BlockID(pid))
 	}
-	return mbk.ToBlock(), nil
+	return bids, nil
 }
 
-// LayerBallots retrieves all ballots from a layer by layer ID.
-func (m *DB) LayerBallots(layerID types.LayerID) ([]*types.Ballot, error) {
-	blocks, err := m.LayerBlocks(layerID)
+// LayerBlocks retrieves all blocks from the layer specified by layer ID.
+func (m *DB) LayerBlocks(layerID types.LayerID) ([]*types.Block, error) {
+	proposals, err := m.LayerProposals(layerID)
 	if err != nil {
-		return nil, fmt.Errorf("layer ballots: %w", err)
+		return nil, err
+	}
+	blocks := make([]*types.Block, 0, len(proposals))
+	for _, p := range proposals {
+		blocks = append(blocks, (*types.Block)(p))
 	}
-	return types.ToBallots(blocks), nil
+	return blocks, nil
+}
+
+// HasProposal returns true if the database has the Proposal specified by the ProposalID and false otherwise.
+func (m *DB) HasProposal(id types.ProposalID) bool {
+	has, err := m.proposals.Has(id.Bytes())
+	return err == nil && has
 }
 
-// LayerBlocks retrieves all blocks from a layer by layer index.
-func (m *DB) LayerBlocks(index types.LayerID) ([]*types.Block, error) {
-	ids, err := m.LayerBlockIds(index)
+// AddProposal adds a proposal to the database.
+func (m *DB) AddProposal(p *types.Proposal) error {
```

Comment:
proposal stores mostly the same data as the ballot. i don't think that duplication is good solution, instead when loading proposal should fetch data from ballots if needed

Reply:
i ended up trimming proposal to the minimum info to store in db.

Reply:
there were two paths to restore a Proposal:
- DBProposal.ToBlock(): used by meshdb clients that are interested in getting Blocks (tortoise/mesh)
- DBProposal.ToProposal(): used by serving peer's request from fetch.GetBlocks(). this is currently how the node sync Proposals/Ballots.